server.js

```diff
@@ -2056,14 +2047,57 @@ if (config.startServer) {
           process.exit(0);
         }
 
-        process.on('SIGTERM', () => {
-          opentelemetry
-            .shutdown()
-            .catch((err) => logger.error('Error shutting down OpenTelemetry', err))
-            .finally(() => {
-              process.exit(0);
-            });
-        });
+        const prepareForTermination = async () => {
+          // We want to proceed with termination even if something goes wrong,
+          // so don't allow this function to throw.
+          try {
+            await util.promisify(server.close).call(server);
+
+            await socketServer.close();
+
+            // Wait for all currently-executing cron jobs to finish.
+            await cron.stop();
+
+            // Wait for all currently-executing server jobs to finish.
+            await serverJobs.stop();
```

Comment:
Did I miss anything else that we should be gracefully stopping? Does this order look correct? As I'm typing this, I'm thinking that maybe we should shut down server jobs before we stop the server and websockets?

Reply:
Ah, external grader results queue processing. I need to implement that.

Reply:
I'm working on this over in https://github.com/PrairieLearn/PrairieLearn/pull/6497.

Reply:
I added external grader results processing shutdown in c0dc2f9.